### ISSUE TITLE
Camera scale affects light intensity

### ISSUE THREAD

**Issue by marwie** [mrdoob/three.js#26659]
### Description

Scaling the camera on spotlight or pointlights affects the light intensity

### Reproduction steps

1. open spotlight or pointlights example
2. change the camera scale
3. see light intensity change

### Code

```js

```

### Live example

https://jsfiddle.net/bmqe9fuy/10/

### Screenshots

https://github.com/mrdoob/three.js/assets/5083203/e16eb760-26fb-4db8-9d30-b0e419f983e7

https://github.com/mrdoob/three.js/assets/5083203/202e4287-a390-4d20-8d42-cf6bc98dfc2f

### Version

154

**Comment by hybridherbst** [comment]
From looking at the code, this seems to be caused by lighting calculations being in view space, but no camera scale being passed in to make sure distance calculations are in world space.

Expectation here would be that camera scale does _not_ affect light intensities. One example where this is necessary is WebXR, where a player becoming larger or smaller must scale the camera as well to get proper IPD and thus proper camera views.
I'm also not aware of other renderers scaling light intensities by camera scale.

I'm not super sure where to add this, but suggestion would be a new cameraScale float uniform on the light state that is then used to make sure the distance calulations are in world space.

**Comment by Mugen87** [comment]
Officially `three.js` does not support scaling the camera. 

Would it be a workaround for WebXR to add the camera to a group and apply the transformation to the group instead?

**Comment by hybridherbst** [comment]
Unfortunately that's not a workaround, as the data is transformed into view space nonetheless, it doesn't really matter where in the chain the transformation is applied.

Good thing is I'm not aware of other areas of three that break with scaled cameras, it's not that uncommon to do that (e.g. some camera control systems work by having the camera parented to an object that's scaled to move the camera further/closer).

I'd be happy to make a PR with some pointers on how to do it right - I started here to minimally invasive pass a cameraScale uniform along the lights data:
https://github.com/needle-tools/three.js/commit/b550ea5216b0ca9ba036fb2c684f001bfbff602b
but am not sure what I have to do to actually update the values besides [the hack here](https://github.com/needle-tools/three.js/commit/b550ea5216b0ca9ba036fb2c684f001bfbff602b#diff-39405de0b68668f8e773fe22299fd39697b084efa55a4b6f2fbe71564da264a4R479).

**Comment by Mugen87** [comment]
Somewhat related: #14372

A PR would be welcome so we can easier discuss the required changes. 👍

**Comment by hybridherbst** [comment]
@Mugen87 I opened a draft PR – my guess is there's just a small piece missing 🤞

**Comment by WestLangley** [comment]
>One example where this is necessary is WebXR, where a player becoming larger or smaller must scale the camera as well to get proper IPD and thus proper camera views.

Perhaps _properly_ setting the IPD is a separate WebXR issue?

>I'm also not aware of other renderers scaling light intensities by camera scale.

Do other renderers support (perspective) camera scale, negative scale, and/or non-uniform scale?

**Comment by hybridherbst** [comment]
@WestLangley as far as I'm aware (and how it works in other engines) scaling the camera is the right approach for WebXR, as WebXR reports views relative to the head transform. WebXR implementations just provide XR views based on that camera transform.

All other viewers I know of support _uniform_ camera scale and lighting stays invariant. So at least Unity, Unreal, Godot, Blender, 3ds Max, Omniverse,... haven't seen one yet that doesn't! I think in applications with visual user input it becomes immediately apparent that something is wrong if one of the 3D views shows different lighting then the others, just because a camera happens to be parented to some object that has scale, or is scaled itself.

Here's a quick overview from tests I did just now:

| | Unity | Blender | Omniverse | Unreal |
| - | - | - | - | - |
| World Scale 1 | ✅ | ✅ | ✅ | ✅ |
| Uniform scale | ✅ | ✅ | ✅ <br/>camera navigation feels awkward | ✅ |
| Non-uniform scale | ✅ <br/> ![image](https://github.com/mrdoob/three.js/assets/2693840/97b15d86-e9ad-4534-be21-31bb646ab92f) | ✅ <br/> ![image](https://github.com/mrdoob/three.js/assets/2693840/06029b12-1763-45a1-bfe1-b49c45319467) | ✅ <br/>lighting is correct<br/>camera navigation feels awkward<br/> ![image](https://github.com/mrdoob/three.js/assets/2693840/01f7e72c-a38b-4147-ac55-d5704af34104) | ✅<br/> ![image](https://github.com/mrdoob/three.js/assets/2693840/f5d49536-1ba2-4486-a0a3-ecd5a64b3419) |
| Negative scale(s) | ✅ <br/>  ![image](https://github.com/mrdoob/three.js/assets/2693840/79ff6634-9048-4ad2-92cc-aadb736c463f) | ⚠️<br/>lighting is correct<br/>viewport is inverted for odd number of flips<br/>shadows look slightly incorrect for odd number of flips<br/> ![20230829-085236_blender-2377](https://github.com/mrdoob/three.js/assets/2693840/ac2ca3a4-d6ce-4a0e-a8d3-b7a055f2f743) | ❌ <br/>rendering breaks for odd number of flips<br/> ![image](https://github.com/mrdoob/three.js/assets/2693840/82a1c0b8-0deb-44c1-9436-3505181a21f7)  | ✅ <br/> ![image](https://github.com/mrdoob/three.js/assets/2693840/65d74580-8e83-4e6e-b44e-7517c04abe9a) |

I'd still welcome some guidance on what the right approach to pass through the uniforms is – at the very least to understand three's rendering better, but I think it would be great to get this issue fixed too.

**Comment by WestLangley** [comment]
The three.js shader code assumes the upper 3x3 of the view matrix is orthogonal -- that is, it's inverse equals its transpose.

https://github.com/mrdoob/three.js/blob/705024703127175c42658778f264aa9b4aa21649/src/renderers/shaders/ShaderChunk/common.glsl.js#L76-L83

(Actually, I think a uniformly-scaled view matrix would also be OK in these cases because the method normalizes the result.)

This method is typically called with the view matrix.

Every use case of `inverseTransformDirection()` in core and the examples would need to be checked.

**Comment by WestLangley** [comment]
The three.js shader code assumes the upper 3x3 of the view matrix is orthogonal -- that is, its inverse equals its transpose.

https://github.com/mrdoob/three.js/blob/705024703127175c42658778f264aa9b4aa21649/src/renderers/shaders/ShaderChunk/common.glsl.js#L76-L83

(Actually, I think a uniformly-scaled view matrix would also be OK in these cases because the method normalizes the result.)

This method is typically called with the view matrix.

Every use case of `inverseTransformDirection()` in core and the examples would need to be checked.

**Comment by WestLangley** [comment]
>Did you have a chance to look at the PR?

Oh yes. I posted my last comment after I read your draft PR.

**Comment by hybridherbst** [comment]
Understood. Could you explain how inverseTransformDirection applies to my PR? (I don't think it does, since it normalizes the result and is not affected by camera scale).

**Comment by WestLangley** [comment]
Correct! As I mentioned, I think the inline comment -- which was written by me -- is overly-restrictive since the result is normalized. So a uniformly-scaled camera seems OK.

What seems weird to me is the idea of using a scaled camera as the way to adjust IPD. But someone other than myself will have to make that call.

**Comment by hybridherbst** [comment]
Just to clarify, I don't mean that per-user IPD differences are adjusted with a scaled camera (they're not) but the optical difference between "I'm an ant" (small distance between eyes) and "I'm a giant" (large distance between eyes) comes from the camera transform in every 3D engine I know (including three.js).

**Comment by hybridherbst** [comment]
To summarize:
- three.js behaviour with scaled cameras differs from all other engines (see table above)
- the camera _transform_ can well be scaled. This scale affects children in the hierarchy, affects WebXR eye distance, and so on. This includes non-linear scaling.
- the camera viewMatrix that is passed to rendering should not contain scale. This does not affect children, WebXR, or anything but rendering. There's a choice to be made regarding behaviour under non-uniform scale with mixed signs; again, see the table above. Unity and Unreal both handle that gracefully (as in: rendering and lighting is unaffected)

**Comment by Mugen87** [comment]
> someone has to make a PR so that viewMatrix is adjusted before being passed on, so scale is ignored

I vote to start with this task in order to match the glTF spec.

**Comment by Mugen87** [comment]
@WestLangley Is it actually possible to remove a non-uniform scale from a matrix? I guess normalizing the basis is not sufficient...

To support scaled ancestor nodes I guess the code would have to go up the hierarchy and compute separate world matrices without scaling.

**Comment by WestLangley** [comment]
@Mugen87 Hmm... A negative scale in x, and a negative scale in y, is equivalent to a rotation in z. So it can be tricky...

A negative determinant is indicative of a reflection.

**Comment by Mugen87** [comment]
A naive approach would be to compute the view matrix by computing additional local matrices of ancestor nodes and the camera without scaling (so use `compose()` with just `position` and `quaternion`), then compute the camera's world matrix and then invert. This code could be placed in `Camera` (maybe `updateViewMatrix()`) and then be used whenever `updateMatrixWorld()` and `updateWorldMatrix()` is called.
